File: microservice-architecture/outbox-pattern/python/test/test_outbox.py
import os
import sys
import unittest

sys.path.append(os.getcwd())

from src.outbox import MemoryBroker, OutboxService  # noqa: E402


class OutboxPatternTest(unittest.TestCase):
    def test_outbox_pattern(self) -> None:
        service = OutboxService()
        broker = MemoryBroker()

        service.create_order("ORD-1001")
        self.assertEqual(1, len(service.orders))
        self.assertEqual("pending", service.outbox[0].status)

        service.relay_pending(broker)
        self.assertEqual(["EVT-ORD-1001"], broker.published)
        self.assertEqual("published", service.outbox[0].status)

        service.relay_pending(broker)
        self.assertEqual(["EVT-ORD-1001"], broker.published)


if __name__ == "__main__":
    unittest.main()

File: microservice-architecture/outbox-pattern/ts/test/test_outbox.ts
import { MemoryBroker, OutboxService } from "../src/outbox";

function assertEqual<T>(expected: T, actual: T, message: string): void {
  if (expected !== actual) {
    throw new Error(`${message} expected=${expected} actual=${actual}`);
  }
}

const service = new OutboxService();
const broker = new MemoryBroker();

service.createOrder("ORD-1001");
assertEqual(1, service.orders.length, "should create order");
assertEqual("pending", service.outbox[0].status, "event should be pending");

service.relayPending(broker);
assertEqual(1, broker.published.length, "event should publish once");
assertEqual("published", service.outbox[0].status, "event should be marked published");

service.relayPending(broker);
assertEqual(1, broker.published.length, "rerun should not duplicate publish");

console.log("outbox-pattern(ts) tests passed");
